### DISCUSSION TITLE
Archive 4.0

### DISCUSSION THREAD

**Discussion by brendan-duncan** [brendan-duncan/archive#237] [Announcements]
The Dart Archive Library is over 9 years old at this point and is showing its age.

I started working on a major overhaul of the library to bring it up to current Dart standards, taking into account that Dart is more heavily used on dart:io platforms than the web these days; to better deal with async, memory, and hopefully performance.

This will be a breaking change, but hopefully it will not be too disruptive. I don't have an estimate release date.

**Comment by brendan-duncan** [comment]
Over a year later and I'm starting work on it again. This started out as such a small library...how did it get so big? Doing a complete overhaul is a lot of work!

**Comment by brendan-duncan** [comment]
@androidseb As someone who has done work with this library, I am working on a breaking change update in the branch https://github.com/brendan-duncan/archive/tree/4.0. If you have any questions or concerns about the changes in 4.0, please let me know. I think it's starting to get to a point where I'm not unhappy with it.

The main emphasis of the 4.0 version is to try and make things more consistent, and to make File IO and reducing memory usage more of a priority.

**Reply by androidseb** [reply]
Thanks for drawing my attention to this, I took a look at the migration guide to get an idea of the changes, these all look like great improvements. Thank you for doing all this work!

I'll try to test this next week and use V4 in my project and see if I encounter any issues, or if I have any feedback for improvements.

**Comment by devoncarew** [comment]
Hi, in case it's useful for the API re-design, here is the info on what parts of this package are used by the top 100 pub.dev packages that reference package:archive:

https://gist.github.com/devoncarew/f02b17041e069cb0f7d20b8034b8684f

Note that this usage info is skewed towards the package use case; typically I'd expect this package to be more used by applications, but those are harder to get concrete data for (package usages are a ~reasonable proxy).

**Comment by brendan-duncan** [comment]
Thanks! Zip seems to be pretty dominant. I'm hoping the general usage will not require too many changes, and hopefully it will make doing the right thing better. And I finally got rid of all those 'dynamic' types! I suspect most usage is for Flutter packages/apps, and now that file streams are more of a first class citizen, I'm hoping people will start using them more for non web apps, get that memory use down.

The github "Used by" list says there are 487,495 dependencies, either direct or indirect!!!! That's crazy!
https://github.com/brendan-duncan/archive/network/dependents

Making breaking changes is always hard, many people will have to update at least something. But I feel like the current version is just getting band-aid after band-aid, and I think it's time to rip the band-aids off and clean it properly. Or, at least as proper as I can manage. I looked at the dart:convert codec classes, I just can't get myself to switch to that API, mostly because it would be a major change and I don't fully understand it. It wouldn't work for archive formats like zip/tar anyways.

**Reply by devoncarew** [reply]
>  I looked at the dart:convert codec classes, I just can't get myself to switch to that API,

Yeah, I wouldn't worry about that - I don't think using that pattern is a necessary part of having a good API.

**Comment by brendan-duncan** [comment]
I started to go through some of the dependent libraries from https://pub.dev/packages?q=dependency%3Aarchive&sort=top to see what would break. I added a few things back to have improved backwards compatibility. Might as well minimize the inconvenience as much as possible.

One of the things people are using is Archive.files. I had changes the design of the Archive class so that it's now hierarchical with ArchiveDirectory and ArchiveFile. I had added some convenience functions getAllEntities (all ArchiveDirectories and ArchiveFiles recursively), and getAllFiles (just the ArchiveFiles recursively). To improve backwards compatibility, I added a files getter back to the Archive class, which returns a cached call of getAllFiles. This made some of the packages I checked now compatible with 4.0 without any changes, like the [lottie](https://pub.dev/packages/lottie) package.

I also added ZLibEncoder.encode, as an alias to the newly named ZLIbEncoder.encodeBytes (done for consistency with decoders). This resolved the backwards incompatibility with the [pdf](https://pub.dev/packages/pdf) package.

Hopefully by adding these types things back in, I can minimize the pain of upgrading. They otherwise have no effect on the new version of the library, other than adding a little extra clutter to the public API.

**Reply by devoncarew** [reply]
Cool; having a glide path from the old to new api makes sense.

The API usage data may or may not be useful for you; it could inform the new api (what parts are heavily used? what parts are never used?). It's useful to know even if if the info isn't immediately actionable.

**Comment by brendan-duncan** [comment]
Looks like I'll need to re-think the Archive + ArchiveFile architecture change I did in 4.0. As pointed out by Brian in https://github.com/brendan-duncan/archive/issues/334, the Dart Build system uses the same ArchiveFile in multiple Archive's. The 3.x architecture allowed for that, but in 4.x, since Archives are hierarchical, ArchiveFile has a parent pointer, and that breaks the idea of having an ArchiveFile in multiple Archive's. Cloning an ArchiveFile when adding it to an Archive has its own complications.

Maybe I'll need to go back to the old design for Archive + ArchiveFile, making them non-hierarchical.

**Comment by androidseb** [comment]
I gave it a quick try but the project doesn't seem to compile for Flutter web for now, I get this compile error:

```
/home/seb/.pub-cache/git/archive-6ae6699dc7576e1957b0bb87db6607d79d0e9107/lib/src/util/input_file_stream.dart:50:30: Error: Too many positional arguments: 0 allowed, but 1 found.
Try removing the extra positional arguments.
        FileBuffer(FileHandle(path), bufferSize: bufferSize),
                             ^
/home/seb/.pub-cache/git/archive-6ae6699dc7576e1957b0bb87db6607d79d0e9107/lib/src/util/_file_handle_html.dart:4:3: Context: Found this candidate, but the arguments don't match.
  FileHandle() {
  ^^^^^^^^^^
/home/seb/.pub-cache/git/archive-6ae6699dc7576e1957b0bb87db6607d79d0e9107/lib/src/util/output_file_stream.dart:37:17: Error: Too many positional arguments: 0 allowed, but 1 found.
Try removing the extra positional arguments.
      FileHandle(path, mode: FileAccess.write),
                ^
```

Since most of my advanced use cases are for the web, my feedback on 4.0 will be limited for now.

The new API design looks good to me, I like all the cleanup and consistency fixes.

For Archive + ArchiveFile being hierarchical, I feel like it is extra functionality that could be wrapped inside separate utility classes rather than baked into the core API: for example, you could have a `HierarchicalArchive` class taking an `Archive` in its constructor, and spitting out `HierarchicalArchiveFile` when queried. This might allow you to better separate the zip serialization code from the convenience file tree navigation utility code and make individual pieces more focused and individually mockable / testable. I haven't looked at the code in depth enough to have a strong opinion on this, so just pointing this out in case it helps.

I'll test this more when I can get the web part to work with RAM based files.

**Reply by brendan-duncan** [reply]
Great, thanks. I'll make sure to check the various Flutters. Probably a class name collision or something, FileHandle is pretty generic. Wish dart had proper namespaces. I did just rewrite the Archive/ArchiveFile classes again, reverted them back to the 3.x style.

**Comment by brendan-duncan** [comment]
@androidseb I fixed the 4.0 branch for html builds, did a simple flutter web test to make sure it didn't complain.

**Comment by androidseb** [comment]
I've tested this a little more, and here is feedback on my personal experience...

The migration to V4 in code is really simple, just need to swap a few function names, the name switches are obvious, IDE auto-completion was enough for me to get through everything, didn't need to look at the migration guide a second time for reference.

I focused my testing on web this time, all the use cases I tested went well (as in no regression), I tested zipping and extraction unencrypted zip files of various sizes
* a few megabytes -> all good
* 500MB -> all good
* 2.4 GB -> all good
* 4.8 GB -> failed like before (can't unzip, didn't test zipping but this is not a regression)

Note that the 4.8GB failure is "expected", this is not a regression, it was already there before. For some reason, in web, I can't read zip files of more than 4GB. I meant to report the issue in the past but I wanted to investigate the root cause a little more and make sure at 100% the issue was coming from the library rather than the way I use it, so take this with a grain of salt. I'll follow up on this separately from the V4 initiative.

My Map Marker app beta is near release (version 3.8) so I won't roll out archive V4 to it just yet. I typically do all the library upgrades at the beginning of the beta phase to give real users time to find and report issues, so I'll first release that 3.8 update, probably in the next couple of weeks, and then I'll issue a new beta version 3.9 with archive v4 in it, and at that time we'll get some real users using archive v4 at scale.

Once I get archive V4 deployed in a beta version, I'll be able to provide performance insights to compare V3 and V4 in terms of speed (my testing was done in debug mode, so not useful to measure performance).

**Comment by brendan-duncan** [comment]
4GB + would need zip64, which should be supported, but I don't think I tested zips that large.

**Reply by androidseb** [reply]
FYI I had tested large zip files 5GB+ on native, zip64 definitely works in systems having dart:io, but for some reason it's not working on web.

Not sure if zip64 decoding is the source of the issue, `RamFileHandle` has a max size of 2GB for memory sub-arrays, and a zip file of 2.4GB can successfully be read, so it should work for a zip file of 4.8GB.

Maybe something about zip64 that has to do with how byte manipulations are done differently in the web version of dart?

**Comment by brendan-duncan** [comment]
The 4.0 branch in now in the main branch. I'll wait a bit before publishing. I had to bite the bullet and move forward at some point, might as well be now.

**Reply by androidseb** [reply]
I got sidetracked with other things and didn't get to testing the V4 yet, I'll work on that in the next few weeks. Can the current commit [0b3ba85fea267a19e8c23796c443cb07c8cb4ecd](https://github.com/brendan-duncan/archive/commit/0b3ba85fea267a19e8c23796c443cb07c8cb4ecd) from the main branch be considered as a "release candidate" for V4?

**Reply by brendan-duncan** [reply]
For the most part. I did just discover a major performance regression with decompressing large files in a zip with file streams, so I will be working on that now. If it takes too long to come up with a solution, I may roll the main branch back to the 3.x commit until I get it sorted out.

For decompressing lots of small files, the 4.0 version is significantly faster and uses less memory. I just have to sort out this cache issue for large files in a zip.

**Reply by brendan-duncan** [reply]
Performance for decompressing zip files on dart:io platforms is much improved now, as it will use native zlib decompression instead of my the Dart decompressor.

**Reply by brendan-duncan** [reply]
The latest commit, a801a1c87a834e668147578206a7cfa36d2cf736, will be considered the release candidate now

**Reply by androidseb** [reply]
I've finally taken some time to test this a little bit.

I noticed that the project main README contains a reference to `archive.getAllEntities` which no longer exists - it should be replaced with `archive.getAllEntities` I think?

I migrated the code and things were smooth as before (for the code part).
However I ran into a bug, it seems the memory is not being released while reading through a zip file, which makes it impossible to read large zip files on mobile.

I modified my code to reproduce the problem in the most straightforward way possible, here is my code:

```dart
  @override
  Future<void> readLargeFile(String largeFilePath) async {
    final InputFileStream inputStream = InputFileStream(largeFilePath);
    final archive = ZipDecoder().decodeStream(inputStream);
    final entities = archive.files;
    int totalBytes = 0;
    for (final entity in entities) {
      if (entity.isSymbolicLink || !entity.isFile) {
        continue;
      }
      final entityBytes = entity.readBytes();
      final bytesCount = entityBytes?.length;
      print('TEST: successfully read file ${entity.name} with ${entityBytes?.length} bytes, total $totalBytes');
      // I remember adding this line back in the days to prevent the memory filling up as I read a zip file, however
      // this time having this "entity.clear()" line active or commented seems to change nothing to the outcome
      await entity.clear();
      if (bytesCount != null) {
        totalBytes += bytesCount;
      }
    }
    print('TEST: finished reading all the files ($totalBytes bytes)');
  }
```

This is the last print I get before the exception happens, so it seems my Android emulator device only has up to 1GB of RAM before the Flutter app reaches max memory:
```
I/flutter ( 6199): TEST: successfully read file local_media/img_9353.jpg with 41612 bytes, total 937332812
```

This is the stack trace:

```
I/flutter ( 6199): Out of Memory
I/flutter ( 6199): #0      ZipFile.getStream (package:archive/src/codecs/zip/zip_file.dart:230:28)
I/flutter ( 6199): #1      ArchiveFile.getContent (package:archive/src/archive/archive_file.dart:139:7)
I/flutter ( 6199): #2      ArchiveFile.readBytes (package:archive/src/archive/archive_file.dart:146:20)
... the rest of the stacktrace comes from where the "entity.readBytes()" call is made in the readLargeFile function
```

I'll try and take more time to further look to see if I can find where this is coming from in the archive library's code.

**Reply by brendan-duncan** [reply]
Thanks for the notes. I'll take a look at the memory with clear.

**Reply by androidseb** [reply]
@brendan-duncan shortly after your last message, I see you pushed some commits to clear the memory in some cases, shall I do another round of testing, or do you need to make more changes first?

**Comment by brendan-duncan** [comment]
I did make some changes that should improve performance in your test. I've been very distracted by work, so I can't quite remember what I pushed. If you want to give it another test, that would be appreciated.

**Reply by androidseb** [reply]
Hello,

I haven't taken a lot of time to test this a lot yet, but I pulled the latest changes and tested again and took a quick look. I can see you fixed the README issues I had pointed out and I can also see that the memory issue I mentioned earlier is gone, I can now complete unzipping a 2.5GB and a 5GB zip file on my 2GB RAM Android emulator.

Something doesn't seem quite right with the end result of the unzipping: after unzipping a 2.5GB file followed by a 5GB file in the Android emulator within the same testing session, it seemed like at least one unzipped entry was missing, but I'm not 100% sure if the library is the root cause, I'll need to test more to confirm and give specifics, I'll try to follow up about this soon.

**Comment by brendan-duncan** [comment]
I haven't had any time to work on this library recently, have a big work milestone coming up and preparing for a conference presentation keeping me really busy.  Hopefully I can get back to polishing the update soon. I haven't seen anything that would cause missing zip entries yet, so I'd want to make sure that doesn't happen.

**Comment by androidseb** [comment]
I tested this more and it seems like the missing unzipped entries issue I encountered earlier was specific to my setup where my test device didn't have enough storage to extract the zip entries, so you can ignore that previously mentioned issue.

I tested again on an Android device with enough storage and I was able to successfully read and write a zip file of 7.5GB.

On native (dart:io) devices, I haven't found any issues testing V4 at the moment.

However, on Web I ran into a problem where reading a RAM-based input stream seems to be not working at all: the read zip files all have 0 bytes.

This is the code I'm using to initiate the InputStream:

```dart
      final ramFileData = RamFileData.fromStream(/* File picker stream params here */);
      return InputFileStream.withFileBuffer(
        FileBuffer(
          RamFileHandle.fromRamFileData(ramFileData),
        ),
      );
```

I ran the library code in debug mode, and I haven't fully understood why this is happening yet, but what I could see is that from this code in `lib/src/codecs/zip/zip_file.dart` (starting at line 228), the call to `decompress.getBytes()` sets `_content` with an empty bytes array:

```dart
//... line 227
      if (compressionMethod == CompressionType.deflate) {
        final savePos = _rawContent!.position;
        final decompress = OutputMemoryStream(size: uncompressedSize);
        ZLibDecoder().decodeStream(_rawContent!, decompress, raw: true);
        _content = decompress.getBytes();
        _rawContent!.setPosition(savePos);
        compressionMethod = CompressionType.none;
      } else if (compressionMethod == CompressionType.bzip2) {
//...
```

I'll keep digging to see if I can find the root cause of this issue.

**Reply by androidseb** [reply]
I took a further look and I have the impression that the issue could come from the differences of implementations of  `_zlib_decoder.dart` between `_zlib_decoder_io.dart` and `_zlib_decoder_web.dart`, but I'm struggling to understand the code to say for sure.

@brendan-duncan I have a few questions about this, maybe you can help me (when you get some time):

1. I have the impression that the encoder / decoders are completely different implementations between io and web, did I get this right?
2. If I got 1. right, what is the main incentive for this decision? Better async support for native platforms maybe?
3. Other than the code, is there any document (outside of the migration guide, which is more of a user-facing guide) I can look at to try and better understand the internal changes of V4?

**Reply by brendan-duncan** [reply]
1. Yes, they are different implementations. For io it uses the native zlib decoder provided by dart:io. Since that is not available on the web, _zlib_decoder_web uses a zlib decoder written in Dart. 

2. Native implementation will obviously have a significant performance benefit, but it's not available on web.

3. Sorry, I don't have anything extra for docs, I'm not that together with this Dart stuff, I can barely keep up with it as it is but I feel too bad about not working on it if it helps people, so I try my best to squeeze in work on it as I can but it means I'm not very organized with it. There were a couple things I had in mind with v4:

a. The previous code evolved from a simple web-only library to do zlib compression and decompression because I needed it for reading and writing PNG images in my other image library. All of the other stuff like zip and tar archives, and dart:io support, were tacked on over time, and the Dart language evolved a lot since it was first written, which was before Dart was even 1.0. There was a lot of use of `dynamic` as a cheat for lack of cleaner design, for example. So I wanted to try and clean up some of that code, and get rid of any use of dynamic.

b. Most of the usage of the library moved from web projects to io projects, particularly since Flutter became a thing (which I haven't even used, other than having to debug a library problem once in a while). Since most of the code was web centric, meaning assuming everything is in memory and the dart:io stuff was more of a tacked on addition, there were a lot of issues with the dart:io centric code, particularly with the buffered input/output streaming, where it was using a lot more memory than it should have, and some issues with doing more io operations than it should, affecting performance. So I tried to fix things up so that the FileStream usage did a better job with those things. I also tried to make sure on dart:io builds that it would use any native provided codecs, like zlib and gzip, if available.

c. I generally rearranged code a bit to be what seemed to me at the time more organized. I also made more of the code available in the archive.dart entry point, for things like the FileStream classes, rather than in archive_io.dart, using conditional imports, as it seemed like a good idea at the time.

I don't do a lot lest testing with the web builds because the testing framework runs with dart:io, so I really do need to add more testing for web builds. I will see if I can add some tests for the RAM stream code stuff, I think that is definitely a weak point in what's being tested, particularly since I didn't write it I don't think of it as much. I'll make sure any issues with that get fixed up.

**Reply by androidseb** [reply]
Thanks for the detailed explanation, it helps.

Personally, I would rather have something consistent and reliable everywhere than something fast only on IO platforms - in V3, I was already quite happy with the performance of the library, but that's probably because I use it as zip library, so I don't mind if the user has to wait a few more seconds to see their import complete, since it's already expected to take some time. Image processing is a separate domain I didn't need, I understand how performance can be more critical there.

Would you consider giving the API to ability to specify an encoder/decoder implementation so that library users who want "the same code to run everywhere" to maximize consistency can do it?

**Comment by androidseb** [comment]
## Archive V4 performance test notes

I tested on my personal Android device running a [Map Marker](https://play.google.com/store/apps/details?id=com.exlyo.mapmarker) Android app production build with the following archive library versions:

* Modified V3 [here](https://github.com/androidseb/archive/commit/94bf62410e9a073451bda0d851a07e8b8863bde4) - that was merged to main but never published as a dedicated version, this is what the production Map Marker app version is currently using
* Latest V4 RC [here](https://github.com/brendan-duncan/archive/commit/a7a0c007299efa0fc9663f6170389d578f5c1a71) - I used my unpublished V4 migration branch to run a production build by side-loading the app

Note that both app versions use Flutter 3.22.1 and roughly the same set of other library dependencies (except the ones that got changed by archive dependencies). The test was done with a 5GB KMZ archive import containing 50000 image files having an average size of 100kB. The Map Marker app import process has the following steps:

1. Initial file copy: copies the archive file into the app storage to be able to manipulate the file directly - this step does not rely on the archive library at all
2. Read archive content: using the archive library to read the zip file and obtain the list of files (without extracting them yet)
3. Extract archive files: using the archive library to extract all files one by one, by copying them to a dedicated internal app folder - files are extracted like this:
	* Extract file data as a byte array (entirely using the archive library)
	* Copy the byte array to the disk (independent from the archive library)
4. Final data processing: checks the imported KML + images files, check that all images from the KML are there, move extracted image files to the user's map directory, etc. - this step does not use the archive library at all

### V3 results

1. Initial file copy completed: 32 - took 32 seconds
2. Archive content read completed: 37 - took 5 seconds
3. Achive decompression completed: 13:07 - took 750 seconds
4. Import completion: 14:20 - took 73 seconds

### V4 results

1. Initial file copy completed: 32 - took 32 seconds
2. Archive content read completed: 41 - took 9 seconds
3. Achive decompression completed: 16:21 - took 940 seconds
4. Import completion: 17:36 - took 75 seconds

**Reply by androidseb** [reply]
Note that I tested manually with a stopwatch, so there could be a 1-2 seconds error margin for each time measurement.

I re-ran steps 1, 2 and the beginning of 3 (the first minute to look at the progress) to ensure the timings were consistent and they seemed to be...

I was surprised to see that the V4 implementation seems to be actually significantly slower by about 25% than V3 from my testing, I did not expect those results.

**Comment by brendan-duncan** [comment]
As usual it's been a while since I've been able to get back to this. I made some more memory and performance fixes. Performance should be much closer to what it was in V3, in some cases faster.

I did a large archive test extracting a 29GB zip, with about 10k jpg and video files of various sizes. 

V3: 
2. Archive content read completed: 1311ms.
3. Decompression (without writing to disk): 123s 

V4:
2. Archive content read completed: 90ms.
3. Decompression (without writing to disk): 108s 

Of course my computer is significantly faster than an Android device, but at least there doesn't seem to be a performance regression anymore.

Here's the code from my test:
```dart
      final t = Stopwatch();
      t.start();
      final input = InputFileStream('d:/photography.zip');
      final archive = ZipDecoder().decodeStream(input);
      t.stop();
      print(t.elapsedMilliseconds);
      t.reset();
      t.start();
      var totalBytes = 0;
      for (final f in archive) {
        if (f.isFile) {
          final bytes = f.readBytes();
          totalBytes += bytes!.length;
          final fp = File('d:/test/${f.name}')..createSync(recursive: true);
          fp.writeAsBytesSync(bytes!);
          f.clear(); // clear up the decompressed memory used by this ArchiveFile
        }
      }
      t.stop();
      print(t.elapsedMilliseconds );
      print('Bytes: $totalBytes');
```

**Reply by androidseb** [reply]
Thanks for the update, I can see you made some fixes to the web implementation, I'll try and clear up some time to test all this again, and I'll post back here when I can. I'm quite busy lately so this could be a little while.

**Comment by brendan-duncan** [comment]
Thanks for having a look whenever you get a chance. No hurry, I'm always busy :-)

One complication about the 4.0 upgrade, but maybe won't be a complication, is build_web_compilers depends on Archive 3.* currently, so I suppose I'll have to let that team know about the update too. I made sure all of the changes are backward compatible with the things they're doing with it.

**Comment by androidseb** [comment]
Hello, I know it's been a while, but better late than never...

Following up on my testing notes from earlier here: https://github.com/brendan-duncan/archive/discussions/237#discussioncomment-10650987

I pulled the latest version from pub.dev (4.0.2 currently) and ran the tests again with it, with the same 5GB file.

* iPad
	* V3 results (using Flutter 2.22)
		* Initial file copy completed: 11s took 11 seconds
    	* Archive content read completed: 11 - took less than a second
    	* Achive decompression completed: 6:18 - took 367 seconds
    	* Import completion: 6:37 - took 19 seconds
	* V4 results (using Flutter 2.27)
    	* Initial file copy completed: 13s - took 13 seconds
    	* Archive content read completed: 13s - took less than a second
    	* Achive decompression completed: 6:13 - took 360 seconds
    	* Import completion: 6:32 - took 19 seconds

* Android device
	* Previous V3 results (using Flutter 2.22, copy/pasted)
    	* Initial file copy completed: 32 - took 32 seconds
    	* Archive content read completed: 37 - took 5 seconds
    	* Achive decompression completed: 13:07 - took 750 seconds
    	* Import completion: 14:20 - took 73 seconds
	* V3 results (using Flutter 2.22, ran again for consistency checking)
		* Initial file copy completed: 32s - took 32 seconds
    	* Archive content read completed: 38s - took 6 seconds
    	* Achive decompression completed: 13:06 - took 748 seconds
    	* Import completion: 14:31 - took 85 seconds
	* V4 results (using Flutter 2.27)
    	* Initial file copy completed: 32s - took 32 seconds
    	* Archive content read completed: 41s - took 9 seconds
    	* Achive decompression completed: 12:33 - took 712 seconds
    	* Import completion: 13:50 - took 73 seconds

In summary, we can see an overall performance improvement from v3 to v4, and I could not find any issues with native or web from my testing so far. Thanks for the great work on this, I'll be sure to report any issues I find in v4.

I hope these notes help :-)
